### PR TITLE
Mac life

### PR DESCRIPTION
**Pull request by outscale-mgo**

### CODE REVIEW

File: src/utils/mac-table.h [hunk 1]
```diff
@@ -53,9 +53,17 @@ struct pg_mac_table {
 };
 
 struct pg_mac_table_iterator {
-	struct pg_mac_table *table;
-	union pg_mac key;
-	int is_end;
+	uint32_t i;
```
Comment (outscale-fne) [inline src/utils/mac-table.h]: Could be great to have a small comment about this variable... I'm not sure what's the purpose of i, the name is not that explicit. Same for 5 next lines... Just to say what is gonna be stored inside!

File: src/utils/mac.h [hunk 2]
```diff
@@ -42,6 +43,8 @@ void pg_get_ether_addrs(const struct rte_mbuf *mb, struct ether_hdr **dest);
 
 void pg_print_mac(const struct ether_addr *eth_addr);
 
+const char *pg_s_printable_mac(const struct ether_addr *eth_addr);
```
Comment (outscale-fne) [inline src/utils/mac.h]: What does s mean inside the naming? Note really sure I understood..

File: src/utils/mac-table.h [hunk 3]
```diff
@@ -137,9 +183,7 @@ static inline void pg_mac_table_elem_set(struct pg_mac_table *ma,
 			pg_mac_table_alloc_fail_exeption(ma);
 	}
 
-	if (pg_mac_table_is_set(*ma->elems[part1], part2))
-		return;
-
+	/* printf("set: %lx\n", mac.mac); */
```
Comment (outscale-toa) [inline src/utils/mac-table.h]: Can you remove comment please?

File: src/utils/mac-table.h [hunk 4]
```diff
@@ -165,30 +209,36 @@ static inline void pg_mac_table_ptr_set(struct pg_mac_table *ma,
...
 {
 	uint32_t part1 = pg_mac_table_part1(mac);
 	uint32_t part2 = pg_mac_table_part2(mac);
 
 	if (!pg_mac_table_is_set(*ma, part1))
-		return;
+		return -1;
 	if (!pg_mac_table_is_set(*ma->ptrs[part1], part2))
-		return;
+		return -1;
```
Comment (outscale-toa) [inline src/utils/mac-table.h]: can you use only one if statement here 
```
if (!pg_mac_table_is_set(*ma, part1) || !pg_mac_table_is_set(*ma->ptrs[part1], part2))
                return -1;
```

File: src/utils/mac-table.h [hunk 5]
```diff
@@ -165,30 +209,36 @@ static inline void pg_mac_table_ptr_set(struct pg_mac_table *ma,
...
+static inline int pg_mac_table_elem_unset(struct pg_mac_table *ma,
+					   union pg_mac mac)
+{
+	uint32_t part1 = pg_mac_table_part1(mac);
+	uint32_t part2 = pg_mac_table_part2(mac);
+
+	if (!pg_mac_table_is_set(*ma, part1))
+		return -1;
+	if (!pg_mac_table_is_set(*ma->elems[part1], part2))
+		return -1;
```
Comment (outscale-toa) [inline src/utils/mac-table.h]: same here

File: src/vtep.c [hunk 6]
```diff
@@ -237,6 +237,17 @@ int pg_vtep_add_mac(struct pg_brick *brick, uint32_t vni,
...
+		      struct ether_addr *mac, struct pg_error **errp)
+{
+	if (!strcmp(pg_brick_type(brick), "vtep4"))
+		return pg_vtep4_unset_mac(brick, vni, mac, errp);
+	else
+		return pg_vtep6_unset_mac(brick, vni, mac, errp);
+
+}
+
+
```
Comment (outscale-fne) [inline src/vtep.c]: Is there two blank lines here? Maybe one should be deleted ?

File: src/utils/mac-table.h [hunk 7]
```diff
@@ -224,102 +274,63 @@ static inline void *pg_mac_table_ptr_get(struct pg_mac_table *ma,
...
+#define MAC_TABLE_TYPE ptr
+#define MAC_TABLE_IT_NEXT pg_mac_table_iterator_next
 
-static inline void pg_mac_table_iterator_init(struct pg_mac_table_iterator *it,
-					      struct pg_mac_table *tbl)
-{
-	it->key.mac = 0;
-	it->is_end = 0;
-	it->table = tbl;
+#include "mac-table-it-next.h"
```
Comment (outscale-toa) [inline src/utils/mac-table.h]: double include of  "mac-table-it-next.h"

File: src/vtep-internal.h [hunk 8]
```diff
@@ -34,9 +34,47 @@ enum operation {
...
+	struct vxlan_hdr vxlan; /* define in rte_ether.h */
+} __attribute__((__packed__));
+
+struct headers_v6 {
+	struct ipv6_hdr	 ip; /* define in rte_ip.h */
+	struct udp_hdr	 udp; /* define in rte_udp.h */
+	struct vxlan_hdr vxlan; /* define in rte_ether.h */
+} __attribute__((__packed__));
+
+
```
Comment (outscale-fne) [inline src/vtep-internal.h]: Two blank lines here... the CI should see this or not? maybe an improvement to make...

File: src/vtep.c [hunk 9]
```diff
@@ -237,6 +237,17 @@ int pg_vtep_add_mac(struct pg_brick *brick, uint32_t vni,
...
 
+int pg_vtep_unset_mac(struct pg_brick *brick, uint32_t vni,
+		      struct ether_addr *mac, struct pg_error **errp)
+{
+	if (!strcmp(pg_brick_type(brick), "vtep4"))
+		return pg_vtep4_unset_mac(brick, vni, mac, errp);
+	else
+		return pg_vtep6_unset_mac(brick, vni, mac, errp);
+
+}
```
Comment (outscale-toa) [inline src/vtep.c]: I think we don't need the "else statement"
```
if (!strcmp(pg_brick_type(brick), "vtep4"))
		return pg_vtep4_unset_mac(brick, vni, mac, errp);
		
return pg_vtep6_unset_mac(brick, vni, mac, errp);
```

File: include/packetgraph/vtep.h [hunk 10]
```diff
@@ -46,6 +46,10 @@ struct ether_addr;
  */
 struct ether_addr *pg_vtep_get_mac(const struct pg_brick *brick);
 
+void pg_vtep4_cleanup_mac(struct pg_brick *brick, int port_id);
+void pg_vtep6_cleanup_mac(struct pg_brick *brick, int port_id);
+
+
```
Comment (outscale-fne) [inline include/packetgraph/vtep.h]: Two empty lines here...

File: tests/vtep/tests.c [hunk 11]
```diff
@@ -1099,93 +1096,169 @@ static void test_vtep_flood_encapsulate(void)
...
+{
+	struct ether_addr mac_src = {{0xb0, 0xb1, 0xb2,
+				       0xb3, 0xb4, 0xb5}};
 	struct pg_error *error = NULL;
-	struct pg_brick *ip_fragment;
-	struct pg_brick *vtep;
-	struct pg_brick *collect_west;
-	struct pg_brick *collect_east;
+	struct ether_addr mac[] = {
+		{{0xc0, 0xc1, 0xc2, 0xc3, 0xc4, 0xc0}},
```
Comment (outscale-fne) [inline tests/vtep/tests.c]: Just a question... Isn't it not better to make a func generating this table? Or would it be worst? (I'm not sure...)

File: include/packetgraph/vtep.h [hunk 12]
```diff
@@ -130,6 +134,13 @@ inline int pg_vtep_add_vni(struct pg_brick *brick,
 int pg_vtep_add_mac(struct pg_brick *brick, uint32_t vni,
 		    struct ether_addr *mac, struct pg_error **errp);
 
+/**
+ * Like what the function name say
+ * Arguments are are the same as pg_vtep_add_mac
+ */
+int pg_vtep_unset_mac(struct pg_brick *brick, uint32_t vni,
+		      struct ether_addr *mac, struct pg_error **errp);
```
Comment (outscale-toa) [inline include/packetgraph/vtep.h]: Add somes descriptions for doxygen documentation
@param description
@param description

File: include/packetgraph/vtep.h [hunk 13]
```diff
@@ -46,6 +46,10 @@ struct ether_addr;
  */
 struct ether_addr *pg_vtep_get_mac(const struct pg_brick *brick);
 
+void pg_vtep4_cleanup_mac(struct pg_brick *brick, int port_id);
+void pg_vtep6_cleanup_mac(struct pg_brick *brick, int port_id);
```
Comment (outscale-toa) [inline include/packetgraph/vtep.h]: Add somes descriptions for doxygen documentation
@param description
@param description

File: src/vtep-internal.c [hunk 14]
```diff
@@ -481,8 +486,8 @@ static inline void add_dst_iner_macs(struct vtep_state *state,
 
 	tmp.mac = 0;
 
-	if (unlikely(port->dead_tables))
-		multicast_mask = pkts_mask;
+	/* if (unlikely(port->dead_tables)) */
+	multicast_mask = vni_mask;
```
Comment (outscale-toa) [inline src/vtep-internal.c]: Can you remove the comment please?

File: src/utils/mac-table.h (hunk 7 above)
Comment (outscale-mgo) [reply src/utils/mac-table.h]: yes

File: src/utils/mac-table.h [hunk 15]
```diff
@@ -41,7 +41,16 @@ struct pg_mac_table_elem {
 };
 
 /**
- * A mac array contening a ptr
+ * A mac array contening pointers or elements
```
Comment (outscale-fne) [inline src/utils/mac-table.h]: *containing

File: src/utils/mac-table.h [hunk 16]
```diff
@@ -41,7 +41,16 @@ struct pg_mac_table_elem {
...
+ * the idea of this mac table, is that a mac is an unique identifier,
+ * as sure, doesn't need hashing we could just
+ * allocate an array for each posible mac
+ * Problem is that doing so require ~280 TByte
+ * So i've cut the mac in 2 part, example 01.02.03.04.05.06
+ * will now have "01.02.03" that will serve as index of the mac table
+ * and "04.05.06" will serve as the index of the submac
+ * if order to take adventage of Virtual Memory, we use bitmask, so we
+ * don't have to allocate 512 MB of physicall ram for each unlucky mac.
```
Comment (outscale-fne) [inline src/utils/mac-table.h]: *physical

File: src/utils/mac-table.h [hunk 17]
```diff
@@ -52,6 +61,14 @@ struct pg_mac_table {
 	jmp_buf *exeption_env;
 };
 
+/**
+ * iterate over a mac table, can't be done using a simple int
+ * as mac table aren't simple array.
+ * a mac table is composed by 1 container of sub container of elements/pointers
+ * i is the index of the main container j, of the current sub container
+ * m0 is current mast of main container and m1 of current sub container
```
Comment (outscale-fne) [inline src/utils/mac-table.h]: Highly likely you wanna say *mask*

File: src/vtep-internal.h [hunk 18]
```diff
@@ -34,9 +34,46 @@ enum operation {
...
+		       struct ether_addr *mac, struct pg_error **errp);
+int pg_vtep6_unset_mac(struct pg_brick *brick, uint32_t vni,
+		       struct ether_addr *mac, struct pg_error **errp);
+
+/*
+ * headers_v4/6 and dest_addresses_v4 are not use in any vtep implementations
+ * but are here for tests
+ */
+struct headers_v4 {
+	struct ipv4_hdr	 ip; /* define in rte_ip.h */
```
Comment (outscale-fne) [inline src/vtep-internal.h]: Maybe *defined... However it's still easily understandable

**Pull request merged**